server/src/main/java/org/opensearch/index/mapper/RootObjectMapper.java

```diff
@@ -301,27 +301,51 @@ public DynamicTemplate[] dynamicTemplates() {
     }
 
     @SuppressWarnings("rawtypes")
-    public Mapper.Builder findTemplateBuilder(ParseContext context, String name, XContentFieldType matchType) {
-        return findTemplateBuilder(context, name, matchType, null);
+    public Mapper.Builder findTemplateBuilder(
+        ParseContext context,
+        String name,
+        XContentFieldType matchType,
+        ObjectMapper.Dynamic dynamic,
+        String fieldFullPath
+    ) {
+        return findTemplateBuilder(context, name, matchType, null, dynamic, fieldFullPath);
     }
 
-    public Mapper.Builder findTemplateBuilder(ParseContext context, String name, DateFormatter dateFormatter) {
-        return findTemplateBuilder(context, name, XContentFieldType.DATE, dateFormatter);
+    public Mapper.Builder findTemplateBuilder(
+        ParseContext context,
+        String name,
+        DateFormatter dateFormatter,
+        ObjectMapper.Dynamic dynamic,
+        String fieldFullPath
+    ) {
+        return findTemplateBuilder(context, name, XContentFieldType.DATE, dateFormatter, dynamic, fieldFullPath);
     }
 
     /**
      * Find a template. Returns {@code null} if no template could be found.
      * @param name        the field name
      * @param matchType   the type of the field in the json document or null if unknown
      * @param dateFormat  a dateformatter to use if the type is a date, null if not a date or is using the default format
+     * @param dynamic     the type of dynamic mapping
+     * @param fieldFullPath the field's full path
      * @return a mapper builder, or null if there is no template for such a field
      */
     @SuppressWarnings("rawtypes")
-    private Mapper.Builder findTemplateBuilder(ParseContext context, String name, XContentFieldType matchType, DateFormatter dateFormat) {
+    public Mapper.Builder findTemplateBuilder(
```

Comment:
I believe we should not deal with `ObjectMapper.Dynamic dynamic` here at all: this is a mapping constraint, not a dynamic template constraint.

Reply:
May be keep the old method and create a new one: `findTemplateBuilderForTemplate`? (since the template is supplied now? Also, please add validation for `dynamicTemplate` (`requireNonNull`)